Added contentWillUnmount prop to call before the Contents unmounts.

Reflex-Gravity opened the pull request:
This PR contains a new feature which adds a new prop `contentWillUnmount` that calls before the Content unmounts.
This PR is linked to #178 .

I have also added tests for this feature.

paztis commented:
The componentWillUnmount poperty mught be optional. Can you add a null test on this.props.componentWillUnmount before execution ?

Reflex-Gravity commented:
> The componentWillUnmount poperty mught be optional. Can you add a null test on this.props.componentWillUnmount before execution ?

I have removed `contentWillUnmount` from the required props, and set the default value of `contentWillUnmount` to null. I have also added tests to support this.

Reflex-Gravity reviewed (commented):
I have removed contentWillUnmount from the required props, and set the default value of contentWillUnmount to null. I have also added tests to support this.

Reflex-Gravity commented:
@paztis Anything else I need to change? Is there something that is blocking from merging?

paztis commented:
No it seems ok. But I'm not owner of the repository and cannot approve it

Le mar. 4 mai 2021 à 10:35, Joel D Souza ***@***.***> a
écrit :

> @paztis <https://github.com/paztis> Anything else I need to change? Is
> there something that is blocking from merging?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ryanseddon/react-frame-component/pull/185#issuecomment-831774604>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIKMMMFD5CL3PSJC5HJA5LTL6WWLANCNFSM433SLBVA>
> .
>

ryanseddon commented on src/Content.jsx:
```diff
@@ -5,7 +5,12 @@ export default class Content extends Component {
     contentDidMount: PropTypes.func.isRequired,
-    contentDidUpdate: PropTypes.func.isRequired
+    contentDidUpdate: PropTypes.func.isRequired,
+    contentWillUnmount: PropTypes.func
```
make this `isRequired` too

ryanseddon commented on src/Content.jsx:
```diff
@@ -16,6 +21,12 @@ export default class Content extends Component {
   }
 
+  componentWillUnmount() {
+    if (typeof this.props.contentWillUnmount === 'function') {
```
Remove the `typeof` check here as you provide a noop in the default props anyway

ryanseddon commented on test/Content.spec.jsx:
```diff
@@ -19,9 +23,14 @@ describe('The Content component', () => {
+      <Content
+        contentDidMount={didMount}
+        contentDidUpdate={didUpdate}
+        contentWillUnmount={willUnmount}
```
leave this off since this isn't testing for `willUnmount` to be called

ryanseddon commented on test/Content.spec.jsx:
```diff
@@ -30,12 +39,17 @@ describe('The Content component', () => {
+      <Content
+        contentDidMount={didMount}
+        contentDidUpdate={didUpdate}
+        contentWillUnmount={willUnmount}
```
Same here remove this

ryanseddon commented on test/Content.spec.jsx:
```diff
@@ -54,4 +68,58 @@ describe('The Content component', () => {
+    expect(didUpdate.callCount).to.equal(0);
+  });
+
+  it('should not error if null is passed in contentWillUnmount', () => {
```
It should error if a user passes null

ryanseddon commented on test/Content.spec.jsx:
```diff
@@ -54,4 +68,58 @@ describe('The Content component', () => {
     });
   });
+
+  it('should call contentWillUnmount before unmounting', () => {
```
This doesn't seem to be testing if `willUnmount` is called?

Reflex-Gravity replied:
I didn't get you.

Reflex-Gravity replied:
Got it.

ryanseddon commented on test/Frame.spec.jsx:
```diff
@@ -313,6 +313,63 @@ describe('The Frame Component', () => {
+    ReactDOM.render(<Parent />, div);
+  });
+
+  it('should not error when null is passed in contentWillUnmount', () => {
```
same as above should error if passed null as prop types expects a `func`

Reflex-Gravity replied:
Can you check the test of "should error"?

ryanseddon reviewed (changes requested):
Looking good just some feedback

ryanseddon commented:
@Reflex-Gravity some feedback for you before we can merge

ryanseddon commented on test/Frame.spec.jsx:
```diff
@@ -313,6 +313,63 @@ describe('The Frame Component', () => {
 
+  it('should call contentWillUnmount on unmount', done => {
+    div = document.body.appendChild(document.createElement('div'));
+    const willUnmount = sinon.spy();
```
I see this is called but this test doesn't check the `callCount` either?

Reflex-Gravity replied:
So, I just need to check the `callCount` to be 1 ryt?

ryanseddon replied:
yep

Reflex-Gravity replied:
I have updated the test, The test is failing, I don't exactly get the reason for this. The implementation of the feature seems to correct and working. But the way I wrote the test is what I guess is wrong. Can you help me here?

ryanseddon replied:
Ok so i think what we're doing here is actually the wrong thing, the issue is about having a callback when the `<Frame>` component unmounts and not when the contents within the iframe unmount.

So I think what we need is a bit simpler instead of passing the `contentWillUnmount` prop down through to `<Content>` we just need to call it inside `<Frame>`'s `componentWillUnmount` callback and we should probably rename it to `frameWillUnmount` to differentiate it from the other content lifecycle events.

Then you can just test it like this:
```js
  it('should call contentWillUnmount on unmount', done => {
    div = document.body.appendChild(document.createElement('div'));
    const willUnmount = sinon.spy();
    const frame = ReactDOM.render(
      <Frame frameWillUnmount={willUnmount} />,
      div
    );
    ReactDOM.unmountComponentAtNode(ReactDOM.findDOMNode(frame).parentNode);
    expect(willUnmount.callCount).to.equal(1, 'expected 1 willUnmount');
    done();
  });
```

Reflex-Gravity replied:
Okay. But when tried it in an example, the `contentWillUnmount` was called when I unmounted the Frame.

ryanseddon replied:
Yeah ok having though on this more I think what you have is useful in two instances of changing the iframe contents as well as unmounting the entire Frame too. 

I honestly have no idea why this isn't firing in test env, if you run `npm run karma:dev` that will allow you to debug the tests within Chrome dev tools.

Reflex-Gravity replied:
Tried with `npm run karma:dev` same issue. How do I proceed? Skip this test?

ryanseddon commented on test/Frame.spec.jsx:
```diff
@@ -317,36 +317,20 @@ describe('The Frame Component', () => {
 
-    wrapper.setState({ isClosed: true }, () => {
+    // TODO: act() doesn'tt seem to solve this something weird???
+    setTimeout(() => {
```
@Reflex-Gravity not sure why `act()` didn't work but this might give you a clue as to what's ahppening here, `srcdoc` is async and `act` doesn't seem to wait long enough for the iframe to actually mount but doing the old `setTimeout` next tick trick does...

ryanseddon commented on test/Frame.spec.jsx:
```diff
@@ -313,6 +313,55 @@ describe('The Frame Component', () => {
+
+  it('should error when null is passed in contentWillUnmount', () => {
+    div = document.body.appendChild(document.createElement('div'));
+    ReactDOM.render(<Frame contentWillUnmount={null} />, div);
```
Also these tests should assert something?

ryanseddon commented on test/Frame.spec.jsx:
```diff
@@ -313,6 +313,55 @@ describe('The Frame Component', () => {
+        }}
+      />,
+      div
+    );
```
Same here nothing is asserted?
